Fix select_dtypes(include='int') for Windows.

OlehKSS opened the pull request:
- [x] closes #36596
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Fix select_dtypes(include='int') for Windows.

Commit:
Whatsnew entry was added.

simonjayhawkins commented on pandas/core/dtypes/common.py:
```diff
@@ -1796,6 +1796,11 @@ def pandas_dtype(dtype) -> DtypeObj:
     # try a numpy dtype
     # raise a consistent TypeError if failed
     try:
+        # int is mapped to different types (int32, in64) on Windows and Linux
```
hmm, unfortunately, so is numpy array construction, 

existing behaviour
```
>>> arr = np.array([1, 2, 3])  # or arr = np.array([1, 2, 3], dtype=int)
>>> arr.dtype
dtype('int32')
>>>
>>> ser = pd.Series(arr)
>>> ser
0    1
1    2
2    3
dtype: int32
>>>
>>> df = pd.DataFrame(ser)
>>>
>>> df.select_dtypes(include="int")
   0
0  1
1  2
2  3
>>>
```

this workflow would break on Windows with this change?

OlehKSS replied:
Yes, it wouldn't work on windows:
```
>>>> df.select_dtypes(include="int")

Empty DataFrame
Columns: []
Index: [0, 1, 2]
```

OlehKSS replied:
In this case, pandas will interpret int as np.int64 and numpy as np.int32 on Windows. Their behavior on Linux will be identical. Though, based on the example in #36596, I see that pandas was mapping integers to np.int64 on Windows by default.

jreback replied:
we don't want to change this routine at all which has far reaching effects. not averse to a tactical change in .select_dtypes itself.

OlehKSS replied:
In that case, I can roll back the changes here and add a warning that `int` is ambiguous and `np.int64` or `np.int32` should be used in `df.select_dtypes` instead.  Another option will be to see why pandas does not follow numpy's approach and treats `int` as `int64` on all platforms.

OlehKSS committed:
Merge branch 'master' into issue-36596

arw2019 commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -331,6 +331,8 @@ Numeric
 - Bug in :class:`Series` where two :class:`Series` each have a :class:`DatetimeIndex` with different timezones having those indexes incorrectly changed when performing arithmetic operations (:issue:`33671`)
 - Bug in :meth:`pd._testing.assert_almost_equal` was incorrect for complex numeric types (:issue:`28235`)
 - Bug in :meth:`DataFrame.__rmatmul__` error handling reporting transposed shapes (:issue:`21581`)
+- Bug in :func:`select_dtypes` different behaviour on windows and linux for ``select_dtypes(include="int")`` (:issue:`36569`)
```
```
- Bug in :func:`select_dtypes` different behavior between Windows and Linux with ``include="int"`` (:issue:`36569`)
```

OlehKSS replied:
Ok

OlehKSS replied:
The changes in this pull request will create problems with the existing pandas behavior, as you can see in the [comment above](https://github.com/pandas-dev/pandas/pull/36808#discussion_r498951543). Do you have any suggestions on how should proceed with this pull request?

arw2019 reviewed (commented):
Small comment. @OlehKSS can you merge master & resolve conflicts?

cc @simonjayhawkins

Commit:
Merge Master

Commit:
Fix code style

Commit:
Merge branch 'issue-36596' of https://github.com/OlehKSS/pandas into issue-36596

Commit:
Fix int inference in select_dtypes.

OlehKSS commented:
@jreback 
What I see as possible changes to `.select_dtypes`:

1. Treat _int_ both as _np.int32_ and _np.int64_. If we have columns with both 64- and 32-bit integers we'll keep all of them
     I implemented this behavior in the current version, so now [this example](https://github.com/pandas-dev/pandas/issues/36596#issue-708024686) and [this exaple](https://github.com/pandas-dev/pandas/pull/36808#discussion_r498951543) should both work
2. Map _int_ to either _np.int32_ or _np.int64_ depending on the type of the given dataframe (_np.int32_, _np.int64_)
3. Raise a warning / error that _int_ is ambiguous, and either _np.int32_ or _np.int64_ should be provided

Another possible routes:
1. Change [maybe_convert_objects](https://github.com/pandas-dev/pandas/blob/d3788525ee368db0b6565b72ab4eeba9140fd1d4/pandas/_libs/lib.pyx#L2115) so it would follow [numpy's behavior](https://github.com/numpy/numpy/issues/9464)
2. Map_int_ to _np.int64_ all the time, as was done previously in this pull request

Commit:
Merge branch 'master' into issue-36596

jbrockmendel commented on pandas/tests/frame/methods/test_select_dtypes.py:
```diff
@@ -82,6 +82,12 @@ def test_select_dtypes_exclude_include_using_list_like(self):
+        include = np.bool_, "int"
+        r = df.select_dtypes(include=include, exclude=exclude)
+        e = df[["b", "e"]]
+        tm.assert_frame_equal(r, e)
```
pls separate this out into a dedicated test with a descriptive name

OlehKSS replied:
Ok

Commit:
Add a separate unit test.

arw2019 reviewed (commented):
@OlehKSS if you merge master and address @jbrockmendel comment we'll re-review

Commit:
Merge branch 'master' into issue-36596

Commit:
Fix dtype in the unit test.

OlehKSS commented:
@arw2019 I merged this branch to the latest master.
@jbrockmendel  I added a separate unit test.
@jreback As you [suggested](https://github.com/pandas-dev/pandas/pull/36808#discussion_r519529870), I moved code from [pandas_dtype](https://github.com/pandas-dev/pandas/blob/4b75067f4f5b01744a9565a21d5ac5057c6a7c43/pandas/core/dtypes/common.py#L1799) into [DataFrame.select_dtypes](https://github.com/pandas-dev/pandas/blob/9769b9dbab018bb192b8776d2db40402b27c388c/pandas/core/frame.py#L3626). In detail, this method will treat _int_ both as _np.int32_ and _np.int64_. Having columns with both 64- and 32-bit integers it will return all of them.
     I implemented this behavior in the current version, so now [this example](https://github.com/pandas-dev/pandas/issues/36596#issue-708024686) and [that exaple](https://github.com/pandas-dev/pandas/pull/36808#discussion_r498951543) should both work.

Commit:
Merge branch 'master' into issue-36596

jreback commented on pandas/tests/frame/methods/test_select_dtypes.py:
```diff
@@ -88,6 +88,29 @@ def test_select_dtypes_exclude_include_using_list_like(self):
+
+        exclude = ("datetime",)
+        include = "bool", int
+        r = df.select_dtypes(include=include, exclude=exclude)
```
call this result and expected

OlehKSS replied:
Ok

jreback commented on pandas/tests/frame/methods/test_select_dtypes.py:
```diff
@@ -88,6 +88,29 @@ def test_select_dtypes_exclude_include_using_list_like(self):
         tm.assert_frame_equal(r, e)
 
+    def test_select_dtypes_exclude_include_int(self):
+        df = DataFrame(
```
can you add the issue number as a comment

OlehKSS replied:
Ok

jreback commented on pandas/tests/frame/methods/test_select_dtypes.py:
```diff
@@ -88,6 +88,29 @@ def test_select_dtypes_exclude_include_using_list_like(self):
+        )
+        exclude = (np.datetime64,)
+        include = np.bool_, "int"
+        r = df.select_dtypes(include=include, exclude=exclude)
```
can you also add the case where inlclude='integer'

OlehKSS replied:
Ok

jreback reviewed (changes requested):
looks ok, can you add a whatsnew note in other enhancements for 1.3 and merge master

OlehKSS committed:
Merge branch 'master' into issue-36596

OlehKSS committed:
Fix whatsnew, add new unit test.

OlehKSS commented:
@jreback I have updated this pull request according to your latest review. Could you take a look at it again?

jreback commented on pandas/tests/frame/methods/test_select_dtypes.py:
```diff
@@ -128,6 +128,36 @@ def test_select_dtypes_exclude_include_using_list_like(self):
+                "f": pd.date_range("now", periods=3).values,
+            }
+        )
+        exclude = (np.datetime64,)
```
can you parameterize on these types (e.g. you can add include & excude as paramters), the this is much simpler to grok.

OlehKSS replied:
Yes, that will be nicer for sure. I updated this pull request with the requested changes.

OlehKSS committed:
Parametrize tests.

OlehKSS committed:
Merge branch 'master' into issue-36596

OlehKSS committed:
Trigger CI.

jreback commented:
thanks @OlehKSS very nice!
